postgis/lwgeom_btree.c

```diff
@@ -131,10 +132,64 @@ PG_FUNCTION_INFO_V1(lwgeom_hash);
 Datum lwgeom_hash(PG_FUNCTION_ARGS)
 {
 	GSERIALIZED *g1 = PG_GETARG_GSERIALIZED_P(0);
+
 	int32_t hval = gserialized_hash(g1);
 	PG_FREE_IF_COPY(g1, 0);
 	PG_RETURN_INT32(hval);
 }
 
+static int
+lwgeom_cmp_abbrev(Datum x, Datum y, SortSupport ssup)
+{
+	/* Empty is a special case */
+	if (x == 0 || y == 0 || x == y)
+		return 0;
+	else if (x > y)
+		return 1;
+	else
+		return -1;
+}
 
+static int
+lwgeom_cmp_full(Datum x, Datum y, SortSupport ssup)
+{
+	GSERIALIZED *g1 = (GSERIALIZED *)PG_DETOAST_DATUM(x);
+	GSERIALIZED *g2 = (GSERIALIZED *)PG_DETOAST_DATUM(y);
+	int ret = gserialized_cmp(g1, g2);
+	return ret;
```

Comment:
Do we need to "free_if_copy" the detoasted datums in the sort support function or the whole context is freed after each comparison?

Reply:
I guess you don't want to waste the tiniest cycle, but the actual implementation of PG_FREE_IF_COPY() just checks if the datum pointer and detoasted pointer are the same, so it's not very expensive to just do-it-all-the-time.

Reply:
Had to pull POSTGIS_FREE_IF_COPY_P from GiST, as there's no such thing in Postgres itself.

Reply:
cmp_full is where we can waste whatever we need, it's a "rare" case of complete equality.